<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Configuration File &mdash; Iliad 1.0 documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../_static/sphinx-argparse.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/js/custom.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Iliad Raw Sequence - Main Module" href="../tutorial/raw_sequence.html" />
    <link rel="prev" title="DEMO" href="demo.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="../index.html">
            <img src="../_static/ILIAD-logo.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Getting started</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="platform_preparation.html">Platform Preparation</a></li>
<li class="toctree-l1"><a class="reference internal" href="demo.html">DEMO</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Configuration File</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">How-To Guides</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../tutorial/raw_sequence.html">Iliad Raw Sequence - Main Module</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorial/stored_sequence.html">Iliad Stored Sequence</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorial/snp_array.html">Iliad SNP Array</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Submodules</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../tutorial/liftover.html">Lift Over Variants - 37 / 38</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorial/merger.html">Merge VCF Data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorial/merge_ref_target.html">Merge Reference and Target Data</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Project Info</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../projectinfo/citing.html">Citing</a></li>
<li class="toctree-l1"><a class="reference internal" href="../projectinfo/more_resources.html">More Resources</a></li>
<li class="toctree-l1"><a class="reference internal" href="../projectinfo/faq.html">Frequently Asked Questions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../projectinfo/contributing.html">Contributing</a></li>
<li class="toctree-l1"><a class="reference internal" href="../projectinfo/contributing.html#types-of-contributions">Types of Contributions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../projectinfo/credits.html">Credits</a></li>
<li class="toctree-l1"><a class="reference internal" href="../projectinfo/changelog.html">Changelog</a></li>
<li class="toctree-l1"><a class="reference internal" href="../projectinfo/license.html">License</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">Iliad</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
      <li>Configuration File</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/getting_started/config.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="configuration-file">
<span id="getting-started-config"></span><h1>Configuration File<a class="headerlink" href="#configuration-file" title="Permalink to this heading">ÔÉÅ</a></h1>
<hr class="docutils" />
<p><strong>The entire contents of the configuration file</strong> located at <code class="docutils literal notranslate"><span class="pre">./Iliad/config/config.yaml</span></code>. Read the comments and subheaders to see the breakdown and explanation of the workflow options.
The most important variables that a user can defined are also stated within each How-To guide.</p>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="c1"># The following configuration file holds very important variables that will help operate I L I A D.</span><span class="w"></span>
<span class="c1"># There are many provided download links that I L I A D will automatically download when you, the user, invoke it.</span><span class="w"></span>
<span class="c1"># All of these come with no warranties. Needless to say, but links tend to break over time!</span><span class="w"></span>
<span class="c1"># We will do our best to keep them up-to-date.</span><span class="w"></span>
<span class="c1"># Feel free to replace any links with your own preferences of files.</span><span class="w"></span>
<span class="c1"># Again, this comes with no warranties.</span><span class="w"></span>

<span class="c1"># __Author__ = Noah Herrick</span><span class="w"></span>
<span class="c1"># __Email__ = noahherrick1@gmail.com</span><span class="w"></span>
<span class="c1"># __Software__ = Iliad: Suite of Snakemake Genomic Data Processing Workflows</span><span class="w"></span>
<span class="c1"># __License__ = MIT License</span><span class="w"></span>
<span class="c1"># __copyright__ = Copyright 2023, Noah Herrick</span><span class="w"></span>
<span class="c1"># __Year__ = 2023</span><span class="w"></span>
<span class="c1"># __Version__ = 1.0.0</span><span class="w"></span>

<span class="c1">#####################################</span><span class="w"></span>
<span class="c1">#####################################</span><span class="w"></span>
<span class="c1">#####################################</span><span class="w"></span>

<span class="c1">#  #  # USER INPUT VARIABLES  #  #  #</span><span class="w"></span>

<span class="c1">#####################################</span><span class="w"></span>
<span class="c1">#####################################</span><span class="w"></span>
<span class="c1">#####################################</span><span class="w"></span>

<span class="c1"># You must insert your /PATH/TO/Iliad/</span><span class="w"></span>
<span class="c1"># use &#39;pwd&#39; command to find your current working directory when you are inside of Iliad directory</span><span class="w"></span>
<span class="c1"># e.g. /user/name/projects/Iliad/ &lt;---- must include forward slash at the end of working directory path</span><span class="w"></span>

<span class="c1"># must include forward slash, &#39;/&#39;, at the end of working directory path</span><span class="w"></span>
<span class="nt">workdirPath</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">NEED PATH HERE</span><span class="w"></span>



<span class="c1">##############################################################################################</span><span class="w"></span>
<span class="c1">### --- Default tables and samples for Raw Sequence (FASTQ) and Stored Sequence (CRAM) --- ###</span><span class="w"></span>
<span class="c1"># ------------------------------------------------------------------------------------------ #</span><span class="w"></span>

<span class="c1"># for downloading FastQ raw seq data make sure there is an Excel table or csv document with two columns and no header: Sample_Name,FTP_url</span><span class="w"></span>
<span class="c1"># e.g KPGP-00127,ftp://ftp.kobic.re.kr/pub/KPGP/2020_release_candidate/WGS_SR/KPGP-00127/KPGP-00127_L1_R1.fq.gz</span><span class="w"></span>
<span class="nt">samplesDict</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">config/UserSampleTable.xlsx</span><span class="w"></span>
<span class="c1"># must include list of samples in one column with &quot;sample&quot; header, no matter if you use the download feature or if you re-direct ILIAD to FASTQ data path</span><span class="w"></span>
<span class="nt">samples</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">config/samples.tsv</span><span class="w"></span>

<span class="c1"># Same setup as above for if you are retrieving CRAM files from an FTP server</span><span class="w"></span>
<span class="nt">cramSamplesDict</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">config/cramSampleTable.xlsx</span><span class="w"></span>
<span class="nt">cramSamples</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">config/cramSamples.tsv</span><span class="w"></span>

<span class="c1">###########################################################</span><span class="w"></span>
<span class="c1">### --- Default GENOME REFERENCE ASSEMBLY retrieval --- ###</span><span class="w"></span>
<span class="c1"># ------------------------------------------------------- #</span><span class="w"></span>

<span class="c1"># If you want to automatically download reference genome assembly, configure below AutoRetrieveReference as &#39;true&#39; - otherwise leave blank!</span><span class="w"></span>
<span class="nt">AutoRetrieveReference</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">true</span><span class="w"> </span><span class="c1"># default is true</span><span class="w"></span>
<span class="c1"># If you already have specific reference genome assembly, configure below IhaveReference as &#39;true&#39;,</span><span class="w"></span>
<span class="c1"># place into your ./Iliad/resources/ directory,</span><span class="w"></span>
<span class="c1"># and configure filename below</span><span class="w"></span>
<span class="c1"># - otherwise leave blank!</span><span class="w"></span>
<span class="nt">IhaveReference</span><span class="p">:</span><span class="w"> </span><span class="c1"># default is blank</span><span class="w"></span>

<span class="c1"># If you have your own reference file to use, state the filePath</span><span class="w"></span>
<span class="c1"># - DO NOT REMOVE &quot;resources/&quot;.</span><span class="w"></span>
<span class="c1"># It MUST be in the your &quot;./Iliad/resources/&quot; directory like so ./Iliad/resources/FILENAME</span><span class="w"></span>
<span class="nt">reference</span><span class="p">:</span><span class="w"></span>
<span class="nt">filePath</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">resources/GRCh38_full_analysis_set_plus_decoy_hla.fa</span><span class="w"> </span><span class="c1"># This is a popular example that you might already have filed away</span><span class="w"></span>

<span class="c1">############################################################</span><span class="w"></span>
<span class="c1">### --- Default VARIANT CALLING options via BCFtools --- ###</span><span class="w"></span>
<span class="c1"># -------------------------------------------------------- #</span><span class="w"></span>

<span class="c1"># BCFtools manual LINK: https://samtools.github.io/bcftools/bcftools.html</span><span class="w"></span>
<span class="c1"># BCFtools cheat sheet LINK: https://gist.github.com/elowy01/93922762e131d7abd3c7e8e166a74a0b</span><span class="w"></span>

<span class="nt">VariantCalling</span><span class="p">:</span><span class="w"></span>
<span class="c1"># # See BCFtools manual for adding additional options, e.g. for base alignment quality &quot;-B&quot;. Just add options within bounds of quotations</span><span class="w"></span>
<span class="nt">mpileup</span><span class="p">:</span><span class="w"></span>
<span class="w">    </span><span class="nt">options</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;-d</span><span class="nv"> </span><span class="s">8000</span><span class="nv"> </span><span class="s">-B&quot;</span><span class="w"> </span><span class="c1"># default is ‚Üí -d 8000 -B</span><span class="w"></span>
<span class="nt">call</span><span class="p">:</span><span class="w"></span>
<span class="w">    </span><span class="nt">options</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;-m</span><span class="nv"> </span><span class="s">-A&quot;</span><span class="w"> </span><span class="c1"># default is ‚Üí -m -A</span><span class="w"></span>

<span class="c1"># Normalize and Left-align - configure below Normalize as &#39;true&#39; - otherwise leave blank!</span><span class="w"></span>
<span class="nt">Normalize</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">true</span><span class="w"></span>

<span class="c1"># # See BCFtools manual for adding additional options. Just add options within bounds of quotations.</span><span class="w"></span>
<span class="c1"># Current options in effect when &quot;Normalize: true&quot; are &quot;norm -f {reference}&quot;</span><span class="w"></span>
<span class="c1"># you can add other flags using the &#39;options: &quot;[add more options here]&quot;&#39; below</span><span class="w"></span>
<span class="nt">Norm</span><span class="p">:</span><span class="w"></span>
<span class="nt">options</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;&quot;</span><span class="w">  </span><span class="c1"># default is blank</span><span class="w"></span>

<span class="c1"># DO NOT Normalize and Left-align - configure below doNotNormalize as &#39;true&#39; - otherwise leave blank!</span><span class="w"></span>
<span class="nt">doNotNormalize</span><span class="p">:</span><span class="w"> </span><span class="c1"># default is blank - benchmarked as true</span><span class="w"></span>

<span class="c1">################################################</span><span class="w"></span>
<span class="c1">### --- Lift and Merge Submodule Options --- ###</span><span class="w"></span>
<span class="c1"># -------------------------------------------- #</span><span class="w"></span>

<span class="c1"># place the appropriate BASE of each filename under the file header &quot;baseFileName_VCF&quot;</span><span class="w"></span>
<span class="c1"># i.e. if FILENAME.vcf, then the BASE is &quot;FILENAME&quot;.</span><span class="w"></span>
<span class="c1"># These can be either compressed (.vcf.gz and .vcf.gz.[tbi/csi]) or uncompressed (.vcf).</span><span class="w"></span>
<span class="c1"># a compressed file will need the associated index file in the directory, too.</span><span class="w"></span>
<span class="nt">vcfs</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">config/mergeTheseVCFs.txt</span><span class="w"></span>

<span class="nt">LiftoverTF</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">true</span><span class="w"> </span><span class="c1"># default is true</span><span class="w"></span>

<span class="c1"># update your genomic positions to Homo sapiens GRCh38 reference assembly - configure below Version38 as &#39;true&#39; - otherwise mark &#39;false&#39;!</span><span class="w"></span>
<span class="nt">Version38</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">true</span><span class="w"> </span><span class="c1"># default is true</span><span class="w"></span>
<span class="c1"># update your genomic positions to Homo sapiens GRCh37 reference assembly - configure above Version38 as &#39;false&#39;</span><span class="w"></span>

<span class="nt">dbsnpLiftMerge</span><span class="p">:</span><span class="w"></span>

<span class="nt">desiredVersion</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">38</span><span class="w"></span>
<span class="nt">projectName</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">Demo</span><span class="w"></span>

<span class="c1">#----------- 37 -------------</span><span class="w"></span>
<span class="nt">dbsnp37VcfDownload</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">https://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh37p13/VCF/All_20180423.vcf.gz</span><span class="w"></span>
<span class="nt">dbsnp37TbiDownload</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">https://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh37p13/VCF/All_20180423.vcf.gz.tbi</span><span class="w"></span>
<span class="nt">file37</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">All_20180423.vcf.gz</span><span class="w"></span>
<span class="c1">#----------- 38 -------------</span><span class="w"></span>
<span class="nt">dbsnp38VcfDownload</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">https://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/VCF/All_20180418.vcf.gz</span><span class="w"></span>
<span class="nt">dbsnp38TbiDownload</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">https://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/VCF/All_20180418.vcf.gz.tbi</span><span class="w"></span>
<span class="nt">file38</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">All_20180418.vcf.gz</span><span class="w"></span>

<span class="nt">genomeReference</span><span class="p">:</span><span class="w"></span>
<span class="c1">#----------- 37 -------------</span><span class="w"></span>
<span class="nt">37Reference</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">http://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.*</span><span class="w"></span>
<span class="c1">#----------- 38 -------------</span><span class="w"></span>
<span class="nt">38Reference</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">http://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/GRCh38_reference_genome/GRCh38_full_analysis_set_plus_decoy_hla.*</span><span class="w"></span>

<span class="c1">#############################################</span><span class="w"></span>
<span class="c1">### --- dbSNP annotation file options --- ###</span><span class="w"></span>
<span class="c1"># ----------------------------------------- #</span><span class="w"></span>

<span class="c1"># used in Raw Sequence Module, Stored Sequence Module, SNP Array Module</span><span class="w"></span>
<span class="c1"># the uncommented configuration options will be used for these modules.</span><span class="w"></span>
<span class="c1"># switch the commented/uncommented three configuration lines if you would like to switch versions.</span><span class="w"></span>
<span class="c1"># you may also update the FTP links if you would like a different dbSNP annotation file - but of course that comes with no warranties</span><span class="w"></span>

<span class="nt">dbSNP</span><span class="p">:</span><span class="w"></span>
<span class="c1"># FTP site: https://ftp.ncbi.nih.gov/snp/</span><span class="w"></span>
<span class="c1"># dbsnp all file</span><span class="w"></span>
<span class="c1"># Check to see if you are using correct assembly with your project</span><span class="w"></span>
<span class="c1">#----------- 37 -------------</span><span class="w"></span>
<span class="c1"># dbsnpVcfDownload: https://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh37p13/VCF/All_20180423.vcf.gz</span><span class="w"></span>
<span class="c1"># dbsnpTbiDownload: https://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh37p13/VCF/All_20180423.vcf.gz.tbi</span><span class="w"></span>
<span class="c1"># file: All_20180423.vcf.gz</span><span class="w"></span>
<span class="c1">#----------- 38 -------------</span><span class="w"></span>
<span class="nt">dbsnpVcfDownload</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">https://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/VCF/All_20180418.vcf.gz</span><span class="w"></span>
<span class="nt">dbsnpTbiDownload</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">https://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/VCF/All_20180418.vcf.gz.tbi</span><span class="w"></span>
<span class="nt">file</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">All_20180418.vcf.gz</span><span class="w"></span>

<span class="c1">#####################################</span><span class="w"></span>
<span class="c1">#####################################</span><span class="w"></span>
<span class="c1">#####################################</span><span class="w"></span>

<span class="c1">#   #   # DEFAULT VARIABLES #   #   #</span><span class="w"></span>

<span class="c1">#####################################</span><span class="w"></span>
<span class="c1">#####################################</span><span class="w"></span>
<span class="c1">#####################################</span><span class="w"></span>

<span class="c1"># used in Raw Sequence Module, Stored Sequence Module, and SNP Array Module</span><span class="w"></span>
<span class="c1"># Reference Genome Assembly</span><span class="w"></span>
<span class="nt">ref</span><span class="p">:</span><span class="w"></span>
<span class="c1"># ensembl species name</span><span class="w"></span>
<span class="nt">species</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">homo_sapiens</span><span class="w"></span>
<span class="c1"># ensembl release</span><span class="w"></span>
<span class="nt">release</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">104</span><span class="w"></span>
<span class="c1"># genome build</span><span class="w"></span>
<span class="nt">build</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">GRCh38</span><span class="w"></span>

<span class="c1"># used in Raw Sequence Module and Stored Sequence Module</span><span class="w"></span>
<span class="c1"># Annotation files for variant calling</span><span class="w"></span>
<span class="nt">NYGC</span><span class="p">:</span><span class="w"></span>
<span class="c1"># FTP Site: http://ftp.1000genomes.ebi.ac.uk/vol1/ftp/data_collections/1000G_2504_high_coverage/working/20190425_NYGC_GATK/</span><span class="w"></span>
<span class="c1"># annotations files are in GRCh38 assembly</span><span class="w"></span>
<span class="nt">nygcUrlPath</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">http://ftp.1000genomes.ebi.ac.uk/vol1/ftp/data_collections/1000G_2504_high_coverage/working/20190425_NYGC_GATK/annotated/</span><span class="w"></span>
<span class="nt">nygcFileStart</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">CCDG_13607_B01_GRM_WGS_2019-02-19_chr</span><span class="w"></span>
<span class="nt">nygcFileEnd</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">.recalibrated_variants.annotated.txt</span><span class="w"></span>
<span class="nt">numberOfSplitRegionsFiles</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">5</span><span class="w"></span>

<span class="c1">###################################</span><span class="w"></span>
<span class="c1">### --- RAW SEQUENCE MODULE --- ###</span><span class="w"></span>
<span class="c1"># ------------------------------- #</span><span class="w"></span>

<span class="c1"># used to shorten the downloaded directories so file is directly placed in correct folder - will need to edit based on your FTP download path</span><span class="w"></span>
<span class="nt">url</span><span class="p">:</span><span class="w"></span>
<span class="nt">cutdirs</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">5</span><span class="w"></span>

<span class="c1">######################################</span><span class="w"></span>
<span class="c1">### --- STORED SEQUENCE MODULE --- ###</span><span class="w"></span>
<span class="c1"># ---------------------------------- #</span><span class="w"></span>

<span class="c1"># used to shorten the downloaded directories so file is directly placed in correct folder - will need to edit based on your FTP download path</span><span class="w"></span>
<span class="nt">cramUrl</span><span class="p">:</span><span class="w"></span>
<span class="nt">cutdirs</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">5</span><span class="w"></span>

<span class="c1">################################</span><span class="w"></span>
<span class="c1">### --- SNP ARRAY MODULE --- ###</span><span class="w"></span>
<span class="c1"># ---------------------------- #</span><span class="w"></span>

<span class="nt">urlProductFiles</span><span class="p">:</span><span class="w"></span>
<span class="c1"># product files LINK</span><span class="w"></span>
<span class="c1"># LINK: https://support.illumina.com/downloads/infinium-multi-ethnic-global-8-v1-product-files.html</span><span class="w"></span>
<span class="c1"># manifest file LINK: update to 37 or 38, make sure its BPM file</span><span class="w"></span>

<span class="c1">#----------- 37 -------------</span><span class="w"></span>
<span class="c1">#manifest: https://webdata.illumina.com/downloads/productfiles/multiethnic-global-8/v1-0/infinium-multi-ethnic-global-8-d1-bpm.zip</span><span class="w"></span>
<span class="c1">#mzip: infinium-multi-ethnic-global-8-d1-bpm.zip</span><span class="w"></span>
<span class="c1">#filename: Multi-EthnicGlobal_D1.bpm # for expanding function later</span><span class="w"></span>
<span class="c1">##build: D1 # for expanding function later</span><span class="w"></span>

<span class="c1">#----------- 38 -------------</span><span class="w"></span>
<span class="c1">## LINK: ftp://ussd-ftp.illumina.com/downloads/productfiles/multiethnic-global-8/v1-0/build38</span><span class="w"></span>
<span class="nt">manifest</span><span class="p">:</span><span class="w">  </span><span class="l l-Scalar l-Scalar-Plain">https://webdata.illumina.com/downloads/productfiles/multiethnic-global-8/v1-0/build38/multi-ethnic-global-8-d2-bpm.zip</span><span class="w"></span>
<span class="nt">mzip</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">multi-ethnic-global-8-d2-bpm.zip</span><span class="w"></span>
<span class="c1">#filename: Multi-EthnicGlobal_D2.bpm # for expanding function later</span><span class="w"></span>
<span class="c1">##build: D2  # for expanding function later</span><span class="w"></span>

<span class="c1"># cluster file LINK</span><span class="w"></span>
<span class="nt">cluster</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">https://webdata.illumina.com/downloads/productfiles/multiethnic-global-8/v1-0/infinium-multi-ethnic-global-8-d1-cluster-file.zip</span><span class="w"></span>
<span class="nt">czip</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">infinium-multi-ethnic-global-8-d1-cluster-file.zip</span><span class="w"></span>

<span class="nt">urlSupportFiles</span><span class="p">:</span><span class="w"></span>
<span class="c1"># support files LINK</span><span class="w"></span>

<span class="c1"># LINK: https://support.illumina.com/downloads/infinium-multi-ethnic-global-8-v1-support-files.html</span><span class="w"></span>

<span class="c1">#----------- 37 ------------- uncomment 37 section below if you need to use GRCh37 assembly</span><span class="w"></span>

<span class="c1"># # physical and genetic coordinates for 37</span><span class="w"></span>
<span class="c1">#physicalGeneticCoordinates: https://support.illumina.com/content/dam/illumina-support/documents/downloads/productfiles/multiethnic-global/multi-ethnic-global-8-d1-physical-genetic-coordinates.zip</span><span class="w"></span>
<span class="c1">#pzip: multi-ethnic-global-8-d1-physical-genetic-coordinates.zip # Multi-EthnicGlobal_D1.csv_Physical-and-Genetic-Coordinates.txt</span><span class="w"></span>

<span class="c1">#----------- 38 ------------- comment block 38 section below if you need to use GRCh37 assembly</span><span class="w"></span>

<span class="c1"># physical and genetic coordinates for 38</span><span class="w"></span>
<span class="nt">physicalGeneticCoordinates</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">https://support.illumina.com/content/dam/illumina-support/documents/downloads/productfiles/multiethnic-global/multi-ethnic-global-8-d2-physical-genetic-coordinates.zip</span><span class="w"></span>
<span class="nt">pzip</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">multi-ethnic-global-8-d2-physical-genetic-coordinates.zip</span><span class="w"> </span><span class="c1"># Multi-EthnicGlobal_D2.csv_Physical-and-Genetic-Coordinates.txt</span><span class="w"></span>

<span class="c1"># rsids conversion file - Loci Name to rsID</span><span class="w"></span>
<span class="nt">rsidConversion</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">https://support.illumina.com/content/dam/illumina-support/documents/downloads/productfiles/multiethnic-global/multi-ethnic-global-8-d2-b150-rsids.zip</span><span class="w"></span>
<span class="nt">rzip</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">multi-ethnic-global-8-d2-b150-rsids.zip</span><span class="w"></span>
<span class="nt">rfile</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">Multi-EthnicGlobal_D2_b150_rsids.txt</span><span class="w"></span>

<span class="nt">Illumina</span><span class="p">:</span><span class="w"></span>
<span class="c1"># iaap-cli exe path</span><span class="w"></span>
<span class="nt">ftpDownload</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">ftp://webdata2:webdata2@ussd-ftp.illumina.com/downloads/software/iaap/iaap-cli-linux-x64-1.1.0.tar.gz</span><span class="w"></span>
<span class="nt">DownloadTarFile</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">iaap-cli-linux-x64-1.1.0.tar.gz</span><span class="w"></span>
<span class="nt">Download</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">iaap-cli-linux-x64-1.1.0</span><span class="w"></span>
<span class="nt">iaapcli</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">iaap-cli</span><span class="w"></span>
<span class="c1">#iaapcli: /N/project/WalshWGS/IliadGenomicDataPipeline/Iliad/target_workflow/illumina_gencall/AutoConvert2.0/AutoConvert</span><span class="w"></span>

<span class="c1">################################</span><span class="w"></span>
<span class="c1">### --- SNP ARRAY MODULE --- ###</span><span class="w"></span>
<span class="c1">### - QC VALUE THRESHOLDS -  ###</span><span class="w"></span>
<span class="c1"># ---------------------------- #</span><span class="w"></span>

<span class="nt">QCarray</span><span class="p">:</span><span class="w"></span>
<span class="w">    </span><span class="nt">GenTrainUpperThreshold</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0.7</span><span class="w"></span>
<span class="w">    </span><span class="nt">GenTrainLowerThreshold</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0.67</span><span class="w"></span>
<span class="w">    </span><span class="nt">ClusterSepUpperThreshold</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0.45</span><span class="w"></span>
<span class="w">    </span><span class="nt">ClusterSepLowerThreshold</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0.4</span><span class="w"></span>


<span class="c1">#####################################</span><span class="w"></span>
<span class="c1">#####################################</span><span class="w"></span>
<span class="c1">#####################################</span><span class="w"></span>

<span class="c1">#  #  #  S U B M O D U L E S  #  #  #</span><span class="w"></span>

<span class="c1">#####################################</span><span class="w"></span>
<span class="c1">#####################################</span><span class="w"></span>
<span class="c1">#####################################</span><span class="w"></span>

<span class="c1"># The major submodule named - Lift-and-Merge - can be found above near line 91.</span><span class="w"></span>
<span class="c1"># There are many configurations, checks, and automatic steps that may help users with little experience.</span><span class="w"></span>
<span class="c1"># These more independent and small task workflows below may come in handy for some quick data maneuvers.</span><span class="w"></span>

<span class="nt">MergerSub</span><span class="p">:</span><span class="w"></span>

<span class="nt">LiftoverSub</span><span class="p">:</span><span class="w"></span>
<span class="c1"># either point to file in config directory or enter 1 filename for file needing converted</span><span class="w"></span>
<span class="c1"># Indicate which reference assembly you desire to switch your positions</span><span class="w"></span>
<span class="nt">filename</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">Tatte-Demo</span><span class="w"></span>
<span class="nt">desiredVersion</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">GRCh38</span><span class="w"> </span><span class="c1"># switch to GRCh37 if you need to revert from 38 to 37</span><span class="w"></span>

<span class="nt">MergeTargetAndRef</span><span class="p">:</span><span class="w"></span>
</pre></div>
</div>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="demo.html" class="btn btn-neutral float-left" title="DEMO" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="../tutorial/raw_sequence.html" class="btn btn-neutral float-right" title="Iliad Raw Sequence - Main Module" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023-2023, Noah Herrick.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>