<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Iliad Raw Sequence - Main Module &mdash; Iliad 1.0 documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../_static/sphinx-argparse.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/js/custom.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Iliad Stored Sequence" href="stored_sequence.html" />
    <link rel="prev" title="Configuration File" href="../getting_started/config.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="../index.html">
            <img src="../_static/ILIAD-logo.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Getting started</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../getting_started/installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../getting_started/platform_preparation.html">Platform Preparation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../getting_started/demo.html">DEMO</a></li>
<li class="toctree-l1"><a class="reference internal" href="../getting_started/config.html">Configuration File</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">How-To Guides</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">Iliad Raw Sequence - Main Module</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#tl-dr-setup">TL;DR setup</a></li>
<li class="toctree-l2"><a class="reference internal" href="#information">Information</a><ul class="simple">
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#background">Background</a></li>
<li class="toctree-l2"><a class="reference internal" href="#basics">Basics</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#in-depth-setup">In-depth Setup</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="stored_sequence.html">Iliad Stored Sequence</a></li>
<li class="toctree-l1"><a class="reference internal" href="snp_array.html">Iliad SNP Array</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Submodules</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="liftover.html">Lift Over Variants - 37 / 38</a></li>
<li class="toctree-l1"><a class="reference internal" href="merger.html">Merge VCF Data</a></li>
<li class="toctree-l1"><a class="reference internal" href="merge_ref_target.html">Merge Reference and Target Data</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Project Info</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../projectinfo/citing.html">Citing</a></li>
<li class="toctree-l1"><a class="reference internal" href="../projectinfo/more_resources.html">More Resources</a></li>
<li class="toctree-l1"><a class="reference internal" href="../projectinfo/faq.html">Frequently Asked Questions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../projectinfo/contributing.html">Contributing</a></li>
<li class="toctree-l1"><a class="reference internal" href="../projectinfo/contributing.html#types-of-contributions">Types of Contributions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../projectinfo/credits.html">Credits</a></li>
<li class="toctree-l1"><a class="reference internal" href="../projectinfo/changelog.html">Changelog</a></li>
<li class="toctree-l1"><a class="reference internal" href="../projectinfo/license.html">License</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">Iliad</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
      <li>Iliad Raw Sequence - Main Module</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/tutorial/raw_sequence.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="iliad-raw-sequence-main-module">
<span id="tutorial-raw-sequence"></span><h1>Iliad Raw Sequence - Main Module<a class="headerlink" href="#iliad-raw-sequence-main-module" title="Permalink to this heading"></a></h1>
<section id="tl-dr-setup">
<h2>TL;DR setup<a class="headerlink" href="#tl-dr-setup" title="Permalink to this heading"></a></h2>
<hr class="docutils" />
<table class="colwidths-given docutils align-center" id="id1">
<caption><span class="caption-text">General Input and Output</span><a class="headerlink" href="#id1" title="Permalink to this table"></a></caption>
<colgroup>
<col style="width: 50%" />
<col style="width: 50%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p>Input</p></td>
<td><p>Output</p></td>
</tr>
<tr class="row-even"><td><p>FASTQ data or FTP links to FASTQ data</p></td>
<td><p>quality-controlled VCF for each chromosome</p></td>
</tr>
</tbody>
</table>
<p><strong>Please make sure that your conda environment for Iliad is activated</strong> - <code class="docutils literal notranslate"><span class="pre">conda</span> <span class="pre">activate</span> <span class="pre">iliadEnv</span></code> or <code class="docutils literal notranslate"><span class="pre">mamba</span> <span class="pre">activate</span> <span class="pre">iliadEnv</span></code></p>
<p><strong>Modify the configuration file</strong> <code class="docutils literal notranslate"><span class="pre">workdirPath</span></code> parameter to the appropriate path leading up to and including <code class="docutils literal notranslate"><span class="pre">/Iliad</span></code> and a final forward slash e.g. <code class="docutils literal notranslate"><span class="pre">/Path/To/Iliad/</span></code>.
The configuration file is found in <code class="docutils literal notranslate"><span class="pre">config/config.yaml</span></code>.</p>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="c1">#####################################</span><span class="w"></span>
<span class="c1">#####################################</span><span class="w"></span>
<span class="c1">#####################################</span><span class="w"></span>

<span class="c1">#  #  # USER INPUT VARIABLES  #  #  #</span><span class="w"></span>

<span class="c1">#####################################</span><span class="w"></span>
<span class="c1">#####################################</span><span class="w"></span>
<span class="c1">#####################################</span><span class="w"></span>

<span class="c1"># You must insert your /PATH/TO/Iliad/</span><span class="w"></span>
<span class="c1"># use &#39;pwd&#39; command to find your current working directory when you are inside of Iliad directory</span><span class="w"></span>
<span class="c1"># e.g. /path/to/Iliad/ &lt;---- must include forward slash at the end of working directory path</span><span class="w"></span>

<span class="c1"># must include forward slash, &#39;/&#39;, at the end of working directory path</span><span class="w"></span>
<span class="nt">workdirPath</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">/Insert/path/to/Iliad/</span><span class="w"></span>
</pre></div>
</div>
<p>You might consider changing some other parameters to your project needs that are pre-set and include:</p>
<ul class="simple">
<li><p>Homo sapiens GRCh38 release 104 reference genome</p></li>
</ul>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">ref</span><span class="p">:</span><span class="w"></span>
<span class="w">  </span><span class="nt">species</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">homo_sapiens</span><span class="w"></span>
<span class="w">  </span><span class="nt">release</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">104</span><span class="w"></span>
<span class="w">  </span><span class="nt">build</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">GRCh38</span><span class="w"></span>
</pre></div>
</div>
<p>Use an Excel sheet or CSV file with no header and the following two columns/fields:</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">Sample   Unique sample identifier</span>
<span class="go">URL   raw sequence data download FTP link</span>
</pre></div>
</div>
<p>Example: <strong>UserSampleTable.xlsx</strong> or <strong>UserSampleTable.csv</strong> are found in the <code class="docutils literal notranslate"><span class="pre">/Iliad/config/</span></code> directory</p>
<table class="colwidths-given docutils align-default" id="id2">
<caption><span class="caption-text">UserSampleTable.xlsx</span><a class="headerlink" href="#id2" title="Permalink to this table"></a></caption>
<colgroup>
<col style="width: 50%" />
<col style="width: 50%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p>KPGP-00127</p></td>
<td><p><a class="reference external" href="ftp://ftp.kobic.re.kr/pub/KPGP/2020_release_candidate/WGS_SR/KPGP-00127/KPGP-00127_L1_R1.fq.gz">ftp://ftp.kobic.re.kr/pub/KPGP/2020_release_candidate/WGS_SR/KPGP-00127/KPGP-00127_L1_R1.fq.gz</a></p></td>
</tr>
<tr class="row-even"><td><p>KPGP-00127</p></td>
<td><p><a class="reference external" href="ftp://ftp.kobic.re.kr/pub/KPGP/2020_release_candidate/WGS_SR/KPGP-00127/KPGP-00127_L1_R2.fq.gz">ftp://ftp.kobic.re.kr/pub/KPGP/2020_release_candidate/WGS_SR/KPGP-00127/KPGP-00127_L1_R2.fq.gz</a></p></td>
</tr>
</tbody>
</table>
<p>This exact template exists already in <code class="docutils literal notranslate"><span class="pre">/Iliad/config/UserSampleTable.xlsx</span></code>. (The Excel Viewer extension on VS code is really handy for editing!)
If you already have the sequence files and are not downloading open-source data, you have the option to place your data into the <code class="docutils literal notranslate"><span class="pre">Iliad/data/fastq/</span></code> directory.</p>
<p>Whether you are automatically downloading via Iliad or you manually place data into <code class="docutils literal notranslate"><span class="pre">Iliad/data/fastq/</span></code> directory,
you need to provide a separate <code class="docutils literal notranslate"><span class="pre">samples.tsv</span></code> file where the TSV file has a header line with only one field named <code class="docutils literal notranslate"><span class="pre">sample</span></code>.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">sample  HEADER</span>
<span class="go">SAMPLE1 sample identifier</span>
<span class="go">SAMPLE2 sample identifier</span>
</pre></div>
</div>
<p>Example: <strong>samples.tsv</strong> found in the <code class="docutils literal notranslate"><span class="pre">/Iliad/config/</span></code> directory</p>
<table class="colwidths-given docutils align-default" id="id3">
<caption><span class="caption-text">samples.tsv</span><a class="headerlink" href="#id3" title="Permalink to this table"></a></caption>
<colgroup>
<col style="width: 100%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p>sample</p></td>
</tr>
<tr class="row-even"><td><p>KPGP-00127</p></td>
</tr>
</tbody>
</table>
<p>since this module is the main snakefile, Snakemake will automatically detect it without the flag.
(Please make sure that your conda environment for Iliad is activated - <code class="docutils literal notranslate"><span class="pre">conda</span> <span class="pre">activate</span> <span class="pre">iliadEnv</span></code> or <code class="docutils literal notranslate"><span class="pre">mamba</span> <span class="pre">activate</span> <span class="pre">iliadEnv</span></code>)</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>snakemake --cores <span class="m">1</span>
</pre></div>
</div>
<p>and combined with other user-specified snakemake flags such as <code class="docutils literal notranslate"><span class="pre">--cores</span></code>.</p>
<p>If you plan to use on a local machine or self-built server without a job scheduler the default command to run is the following:</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>snakemake -p --use-singularity --use-conda --cores <span class="m">1</span> --jobs <span class="m">1</span> --default-resource<span class="o">=</span><span class="nv">mem_mb</span><span class="o">=</span><span class="m">10000</span> --latency-wait <span class="m">120</span>
</pre></div>
</div>
<p>However, there is a file included in the <code class="docutils literal notranslate"><span class="pre">Iliad</span></code> directory named - <code class="docutils literal notranslate"><span class="pre">snakemake.sh</span></code> that will be useful in batch job submission.
Below is an example snakemake workflow submission in SLURM job scheduler.
Please read the shell variables at the top of the script and customize to your own paths and resource needs.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>sbatch snakemake.sh
</pre></div>
</div>
<p>If you would like more in-depth information and descriptions, please continue to the next sections below.
Otherwise, you have completed the TL;DR setup section.</p>
</section>
<section id="information">
<h2>Information<a class="headerlink" href="#information" title="Permalink to this heading"></a></h2>
<hr class="docutils" />
<p>This How-To Guide introduces the raw sequence read data processing module of the <a href="#id5"><span class="problematic" id="id6">Iliad_</span></a> workflow developed using Snakemake workflow language.
Please visit <a class="reference external" href="https://snakemake.readthedocs.io">Snakemake</a> for specific details. They also provide informational <a class="reference external" href="https://slides.com/johanneskoester/snakemake-tutorial">slides</a>. In general, though, each module is composed of rules. These rules define how output files are generated from input files while
automatically determining dependencies amongst the rules. A <code class="docutils literal notranslate"><span class="pre">DAG</span></code> (directed acyclic graph) of jobs will be built each time to account for all the samples and jobs
that will be executed either via a job scheduler or local cores and will execute in parallel if multiple jobs are declared.
Because of the Snakemake workflow system design, the <strong>Iliad</strong> workflow is scalable from single core machines to HPC clusters with job schedulers.</p>
<p>The <strong>Raw Sequence Module</strong> is designed to process FASTQ sequence files, whether they are single-end or paired-end reads.
Currently, it is most common to prepare <strong>Reference Data</strong> Panels from open-source Whole-Genome Sequence (WGS) data such as 1000 Genomes Project <a class="reference external" href="http://ftp.1000genomes.ebi.ac.uk/vol1/ftp/">data</a>.
Reference data is paramount in preparation of target data samples, especially in human research studies where biogeographical ancestry is either the variable of interest
or covariate, for instance in GWAS.
Depending on the level of data sharing, some research labs are unable to submit their target data to commercial imputation services and will need to build their own <strong>Reference Data</strong> Panel
as well as depend on open-source Imputation tools. We highly recommend <a class="reference external" href="https://doi.org/10.1186/s12859-019-2964-5">Odyssey</a>: <cite>a semi-automated pipeline for phasing, imputation, and analysis of genome-wide genetic data</cite>.</p>
<p>As sequence data gradually overtakes microarray data as the primary <strong>Target Data</strong> source, it is going to be mission critical
to derive research related genotypic information quickly and efficiently from raw sequence data, too.
We ensured no bioinformatics knowledge is needed to run this module with the help of external test runs performed on Google Cloud Platform (<a class="reference external" href="https://cloud.google.com/">GCP</a>).</p>
<p><strong>Raw Sequence Module Rule Graph</strong></p>
<a class="reference internal image-reference" href="../_images/raw-seq-rulegraph.png"><img alt="../_images/raw-seq-rulegraph.png" class="align-center" src="../_images/raw-seq-rulegraph.png" style="width: 400px;" /></a>
<div class="toctree-wrapper compound">
</div>
</section>
<section id="background">
<h2>Background<a class="headerlink" href="#background" title="Permalink to this heading"></a></h2>
<hr class="docutils" />
<p>Sequencing data is becoming more accessible for researchers. It’s primary role for years in GWAS was to serve as reference data to fill in the gaps of
target datasets containing large sample sizes through phasing and imputation techniques. This will continue to be a very important role as microarray data is
still commonly used and more sequence data means bigger and better reference sets for this purpose. Sequence data’s current role, however, is already in transition for
its use as target data. The automation of the numerous steps to prepare data from a sequencer’s raw output is crucial, so we developed <cite>Iliad</cite> with users of all levels
in mind to obtain variant call files with clean and reliable genotypes. It does possess the capability to be adapted and have add-ons.
Pull requests and contributions are welcomed.</p>
</section>
<section id="basics">
<h2>Basics<a class="headerlink" href="#basics" title="Permalink to this heading"></a></h2>
<hr class="docutils" />
<p>The raw files from a next-generation sequencer are FASTQ files found in <code class="docutils literal notranslate"><span class="pre">.fastq</span></code> or <code class="docutils literal notranslate"><span class="pre">.fq</span></code> format, and they are often zipped <code class="docutils literal notranslate"><span class="pre">fastq.gz</span></code> or <code class="docutils literal notranslate"><span class="pre">fq.gz</span></code>.
There are instances where the data may be single-end reads or paired-end reads. In the case of <strong>paired-end reads</strong>, sample naming schemes can and will vary,
but will have some form of <code class="docutils literal notranslate"><span class="pre">Sample1_R1.fq.gz</span></code> and <code class="docutils literal notranslate"><span class="pre">Sample1_R2.fq.gz</span></code>. Paired-end reads are sequenced from both ends of the specified DNA read length.
<cite>Iliad</cite> raw sequence data module was designed to handle both types of sequence data but was tested using paired-end reads.
As the user, you have the choice to upload an Excel sheet or CSV file with no header and the following two columns/fields:</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">Sample   Unique sample identifier</span>
<span class="go">URL   raw sequence data download FTP link</span>
</pre></div>
</div>
<p>If you already have the sequence files and are not downloading open-source data, you have the option to place your data into the <code class="docutils literal notranslate"><span class="pre">Iliad/data/fastq/</span></code> directory.
You will still need to provide a separate <code class="docutils literal notranslate"><span class="pre">samples.tsv</span></code> file in both situations where the TSV file has a header line with only one field named <code class="docutils literal notranslate"><span class="pre">sample</span></code>.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">sample  HEADER</span>
<span class="go">SAMPLE1 sample identifier</span>
<span class="go">SAMPLE2 sample identifier</span>
</pre></div>
</div>
<p>These <code class="docutils literal notranslate"><span class="pre">.fq.gz</span></code> files will undergo multiplex concatenation if needed and sent to BWA mem alignment and sorting to create a sorted.bam file and it’s index.
This requires a reference genome assembly and Iliad downloads the user-configured reference genome fasta files.
Iliad is configured to download <em>Homo sapiens</em> GRCh38 release 104 as default.</p>
<p><a class="reference external" href="https://samtools.github.io/bcftools/bcftools.html">BCFtools</a> is used to perform mpileup and call to perform variant calling on New York Genome Center (NYGC) annotated sites. <cite>Iliad</cite> downloads and uses the NYGC’s
annotated SNP regions file per chromosomes (link to <a class="reference external" href="http://ftp.1000genomes.ebi.ac.uk/vol1/ftp/data_collections/1000G_2504_high_coverage/working/20190425_NYGC_GATK/annotated/CCDG_13607_B01_GRM_WGS_2019-02-19_chr22.recalibrated_variants.annotated.txt">chr22</a> file) - <cite>recalibrated_variants.annotated.txt</cite>.
A README file can be found concerning the annotations <a class="reference external" href="http://ftp.1000genomes.ebi.ac.uk/vol1/ftp/data_collections/1000G_2504_high_coverage/working/20190425_NYGC_GATK/1000G_README_2019April10_NYGCjointcalls.pdf">protocol</a>.
These chromosome annotations files consume roughly 70 GB of storage, so we process the files into a simple regions file with two fields and no header: CHROM POS.
This reduces the file footprint to a little over 1 GB. The limitation of this file is <cite>Homo sapiens</cite> GRCh38 reference assembly.
There are 124,954,769 variants to be called across chromosomes 1-22 and X.</p>
<p>This is a resource intensive task to be performed on many samples and so we reduce the serial runtime by splitting each chromosome regions file into even splits
that can be processed in parallel with job scheduling and management controlled by Snakemake. We can then concatenate the splits per chromosome per sample to obtain 23
chromosome VCFs for each sample. Samples are merged at each chromosome and the ID field in the VCF is annotated using BCFtools and dbSNP annotations <a class="reference external" href="https://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/VCF/">vcf</a> so that
each position is given a standardized <code class="docutils literal notranslate"><span class="pre">rs</span> <span class="pre">ID</span></code>. The default configuration file is set to download human_9606_b151_GRCh38p7 <code class="docutils literal notranslate"><span class="pre">All_20180418.vcf.gz</span></code>.</p>
<section id="in-depth-setup">
<h3>In-depth Setup<a class="headerlink" href="#in-depth-setup" title="Permalink to this heading"></a></h3>
<p>Once the <a class="reference external" href="https://iliad-readthedocs.readthedocs.io/en/latest/getting_started/installation.html">Installation</a> of Iliad and its two dependencies has been completed,
you will find your new working directory within the <code class="docutils literal notranslate"><span class="pre">PATH/TO/Iliad</span></code> folder.
Make sure your current working directory is in this cloned repo as stated in the installation.
If the repository is not cloned in that fashion, there is a chance that your directory will be improperly named as <code class="docutils literal notranslate"><span class="pre">Iliad-main</span></code>.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span><span class="nb">cd</span> Iliad
</pre></div>
</div>
<p>In that working directory you will find there are a number of directories with files and code to run each of the module pipelines.</p>
<p><strong>FIRST</strong>,
there is a configuration file with some default parameters, however, you MUST at least change the <code class="docutils literal notranslate"><span class="pre">workdirPath</span></code> parameter to the appropriate
path leading up to and including <code class="docutils literal notranslate"><span class="pre">/Iliad</span></code> e.g. <code class="docutils literal notranslate"><span class="pre">/Path/To/Iliad/</span></code>. The configuration file is found in <code class="docutils literal notranslate"><span class="pre">config/config.yaml</span></code>.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">workdirPath</span><span class="p">:</span> <span class="o">/</span><span class="n">Path</span><span class="o">/</span><span class="n">To</span><span class="o">/</span><span class="n">Iliad</span><span class="o">/</span>
</pre></div>
</div>
<p>You might consider changing some other parameters to your project needs that are pre-set and include:</p>
<ul class="simple">
<li><p>Homo sapiens GRCh38 release 104 reference genome</p></li>
</ul>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">ref</span><span class="p">:</span>
  <span class="n">species</span><span class="p">:</span> <span class="n">homo_sapiens</span>
  <span class="n">release</span><span class="p">:</span> <span class="mi">104</span>
  <span class="n">build</span><span class="p">:</span> <span class="n">GRCh38</span>
</pre></div>
</div>
<ul class="simple">
<li><p>URL and path to FTP site housing the data for retrieval. Here is an example from</p></li>
</ul>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">url</span><span class="p">:</span>
   <span class="c1"># number of directories that the desired file is in at ftp site</span>
   <span class="c1"># for example</span>
   <span class="c1"># ftp://ftp.site.host.example/path/to/data/files/</span>
   <span class="n">cutdirs</span><span class="p">:</span> <span class="mi">5</span>
</pre></div>
</div>
<p><strong>SECOND</strong>,
there is a <code class="docutils literal notranslate"><span class="pre">data/fastq</span></code> directory with a <code class="docutils literal notranslate"><span class="pre">readme.md</span></code> file. You must place all of your <code class="docutils literal notranslate"><span class="pre">.fq.gz</span></code> files in this folder <strong>IF you have the files already</strong>.
Otherwise, as mentioned above, <cite>Iliad</cite> features a downloading step, particularly from FTP sites where open-source genomic data is hosted.
You will need to provide a table of either Excel or CSV format as seen below. There is no header line so simply replace Sample with your unique identifier for your sample and
URL for the specific URL path to the FTP site that hosts the data.</p>
<table class="colwidths-given docutils align-default" id="id4">
<caption><span class="caption-text">UserSampleTable.xlsx or UserSampleTable.csv are found in the <code class="docutils literal notranslate"><span class="pre">/Iliad/config/</span></code> directory</span><a class="headerlink" href="#id4" title="Permalink to this table"></a></caption>
<colgroup>
<col style="width: 50%" />
<col style="width: 50%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p>Sample</p></td>
<td><p>URL</p></td>
</tr>
</tbody>
</table>
<p>You’ll notice in provided templates, there are multiple rows of the same sample. Looking closely, you will find that these samples are multiplexed e.g. same samples,
but sequenced multiple times on different lanes (L1, L2, etc.), and this is perfectly okay.</p>
<p><strong>THIRD</strong>,
each module pipeline has a specific <code class="docutils literal notranslate"><span class="pre">Snakefile</span></code>.
Snakemake will automatically detect the main snakefile, which is named excatly as such and found in the <code class="docutils literal notranslate"><span class="pre">workflow</span></code> directory: <code class="docutils literal notranslate"><span class="pre">workflow/Snakefile</span></code>.
Iliad reserves the main snakefile for the main module, specifically this raw sequence read data <a class="reference external" href="https://iliad-readthedocs.readthedocs.io/en/latest/tutorial/raw_sequence.html">module</a>.
This means the user must specify which <code class="docutils literal notranslate"><span class="pre">Snakefile</span></code> will be invoked with the following:</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>snakemake --snakefile workflow/Snakefile
</pre></div>
</div>
<p><strong>OR</strong> since this module is the main snakefile, Snakemake will automatically detect it without the flag.
(Please make sure that your conda environment for Iliad is activated - <code class="docutils literal notranslate"><span class="pre">conda</span> <span class="pre">activate</span> <span class="pre">IliadEnv</span></code> or <code class="docutils literal notranslate"><span class="pre">mamba</span> <span class="pre">activate</span> <span class="pre">IliadEnv</span></code>)</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>snakemake --cores <span class="m">1</span>
</pre></div>
</div>
<p>and combined with other user-specified snakemake flags such as <code class="docutils literal notranslate"><span class="pre">--cores</span></code>.</p>
<p>If you plan to use on a local machine or self-built server without a job scheduler the default command to run is the following:</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>snakemake -p --use-singularity --use-conda --cores <span class="m">1</span> --jobs <span class="m">1</span> --default-resource<span class="o">=</span><span class="nv">mem_mb</span><span class="o">=</span><span class="m">10000</span> --latency-wait <span class="m">120</span>
</pre></div>
</div>
<p>However, there is a file included in the <code class="docutils literal notranslate"><span class="pre">Iliad</span></code> directory named - <code class="docutils literal notranslate"><span class="pre">snakemake.sh</span></code> that will be useful in batch job submission.
Below is an example snakemake workflow submission in SLURM job scheduler.
Please read the shell variables at the top of the script and customize to your own paths and resource needs.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>sbatch snakemake.sh
</pre></div>
</div>
</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="../getting_started/config.html" class="btn btn-neutral float-left" title="Configuration File" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="stored_sequence.html" class="btn btn-neutral float-right" title="Iliad Stored Sequence" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023-2023, Noah Herrick.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>