<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Iliad SNP Array &mdash; Iliad 1.0 documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../_static/sphinx-argparse.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/js/custom.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Lift Over Variants - 37 / 38" href="liftover.html" />
    <link rel="prev" title="Iliad Stored Sequence" href="stored_sequence.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="../index.html">
            <img src="../_static/ILIAD-logo.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Getting started</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../getting_started/installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../getting_started/platform_preparation.html">Platform Preparation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../getting_started/demo.html">DEMO</a></li>
<li class="toctree-l1"><a class="reference internal" href="../getting_started/config.html">Configuration File</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">How-To Guides</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="raw_sequence.html">Iliad Raw Sequence - Main Module</a></li>
<li class="toctree-l1"><a class="reference internal" href="stored_sequence.html">Iliad Stored Sequence</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Iliad SNP Array</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#tl-dr-setup">TL;DR setup</a></li>
<li class="toctree-l2"><a class="reference internal" href="#information">Information</a><ul class="simple">
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#background">Background</a></li>
<li class="toctree-l2"><a class="reference internal" href="#basics">Basics</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#in-depth-setup">In-depth Setup</a></li>
</ul>
</li>
</ul>
</li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Submodules</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="liftover.html">Lift Over Variants - 37 / 38</a></li>
<li class="toctree-l1"><a class="reference internal" href="merger.html">Merge VCF Data</a></li>
<li class="toctree-l1"><a class="reference internal" href="merge_ref_target.html">Merge Reference and Target Data</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Project Info</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../projectinfo/citing.html">Citing</a></li>
<li class="toctree-l1"><a class="reference internal" href="../projectinfo/more_resources.html">More Resources</a></li>
<li class="toctree-l1"><a class="reference internal" href="../projectinfo/faq.html">Frequently Asked Questions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../projectinfo/contributing.html">Contributing</a></li>
<li class="toctree-l1"><a class="reference internal" href="../projectinfo/contributing.html#types-of-contributions">Types of Contributions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../projectinfo/credits.html">Credits</a></li>
<li class="toctree-l1"><a class="reference internal" href="../projectinfo/changelog.html">Changelog</a></li>
<li class="toctree-l1"><a class="reference internal" href="../projectinfo/license.html">License</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">Iliad</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
      <li>Iliad SNP Array</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/tutorial/snp_array.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="iliad-snp-array">
<span id="tutorial-snp-array"></span><h1>Iliad SNP Array<a class="headerlink" href="#iliad-snp-array" title="Permalink to this heading"></a></h1>
<section id="tl-dr-setup">
<h2>TL;DR setup<a class="headerlink" href="#tl-dr-setup" title="Permalink to this heading"></a></h2>
<hr class="docutils" />
<table class="colwidths-given docutils align-center" id="id1">
<caption><span class="caption-text">General Input and Output</span><a class="headerlink" href="#id1" title="Permalink to this table"></a></caption>
<colgroup>
<col style="width: 50%" />
<col style="width: 50%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p>Input</p></td>
<td><p>Output</p></td>
</tr>
<tr class="row-even"><td><p>IDAT data</p></td>
<td><p>quality-controlled VCF</p></td>
</tr>
</tbody>
</table>
<p><strong>Please make sure that your conda environment for Iliad is activated</strong> - <code class="docutils literal notranslate"><span class="pre">conda</span> <span class="pre">activate</span> <span class="pre">iliadEnv</span></code> or <code class="docutils literal notranslate"><span class="pre">mamba</span> <span class="pre">activate</span> <span class="pre">iliadEnv</span></code></p>
<p><strong>Modify the configuration file</strong> <code class="docutils literal notranslate"><span class="pre">workdirPath</span></code> parameter to the appropriate path leading up to and including <code class="docutils literal notranslate"><span class="pre">/Iliad</span></code> and a final forward slash e.g. <code class="docutils literal notranslate"><span class="pre">/Path/To/Iliad/</span></code>.
The configuration file is found in <code class="docutils literal notranslate"><span class="pre">config/config.yaml</span></code>.</p>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="c1">#####################################</span><span class="w"></span>
<span class="c1">#####################################</span><span class="w"></span>
<span class="c1">#####################################</span><span class="w"></span>

<span class="c1">#  #  # USER INPUT VARIABLES  #  #  #</span><span class="w"></span>

<span class="c1">#####################################</span><span class="w"></span>
<span class="c1">#####################################</span><span class="w"></span>
<span class="c1">#####################################</span><span class="w"></span>

<span class="c1"># You must insert your /PATH/TO/Iliad/</span><span class="w"></span>
<span class="c1"># use &#39;pwd&#39; command to find your current working directory when you are inside of Iliad directory</span><span class="w"></span>
<span class="c1"># e.g. /path/to/Iliad/ &lt;---- must include forward slash at the end of working directory path</span><span class="w"></span>

<span class="c1"># must include forward slash, &#39;/&#39;, at the end of working directory path</span><span class="w"></span>
<span class="nt">workdirPath</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">/Insert/path/to/Iliad/</span><span class="w"></span>
</pre></div>
</div>
<p>You might consider changing some other parameters to your project needs that are pre-set and include:</p>
<ul class="simple">
<li><p>Homo sapiens GRCh38 release 104 reference genome</p></li>
</ul>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">ref</span><span class="p">:</span><span class="w"></span>
<span class="w">  </span><span class="nt">species</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">homo_sapiens</span><span class="w"></span>
<span class="w">  </span><span class="nt">release</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">104</span><span class="w"></span>
<span class="w">  </span><span class="nt">build</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">GRCh38</span><span class="w"></span>
</pre></div>
</div>
<ul class="simple">
<li><p>Illumina <a class="reference external" href="https://support.illumina.com/array/array_kits/infinium-multi-ethnic-global-8-kit.html">MEGA</a> microarray GRCh38 support and product files</p></li>
</ul>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">urlProductFiles</span><span class="p">:</span><span class="w"></span>
<span class="w">  </span><span class="nt">manifest</span><span class="p">:</span><span class="w">  </span><span class="l l-Scalar l-Scalar-Plain">https://webdata.illumina.com/downloads/productfiles/multiethnic-global-8/v1-0/build38/multi-ethnic-global-8-d2-bpm.zip</span><span class="w"></span>
<span class="w">  </span><span class="nt">mzip</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">multi-ethnic-global-8-d2-bpm.zip</span><span class="w"></span>
<span class="w">  </span><span class="nt">cluster</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">https://webdata.illumina.com/downloads/productfiles/multiethnic-global-8/v1-0/infinium-multi-ethnic-global-8-d1-cluster-file.zip</span><span class="w"></span>
<span class="w">  </span><span class="nt">czip</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">infinium-multi-ethnic-global-8-d1-cluster-file.zip</span><span class="w"></span>

<span class="nt">urlSupportFiles</span><span class="p">:</span><span class="w"></span>
<span class="w">  </span><span class="nt">physicalGeneticCoordinates</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">https://support.illumina.com/content/dam/illumina-support/documents/downloads/productfiles/multiethnic-global/multi-ethnic-global-8-d2-physical-genetic-coordinates.zip</span><span class="w"></span>
<span class="w">  </span><span class="nt">pzip</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">multi-ethnic-global-8-d2-physical-genetic-coordinates.zip</span><span class="w"></span>
<span class="w">  </span><span class="nt">rsidConversion</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">https://support.illumina.com/content/dam/illumina-support/documents/downloads/productfiles/multiethnic-global/multi-ethnic-global-8-d2-b150-rsids.zip</span><span class="w"></span>
<span class="w">  </span><span class="nt">rzip</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">multi-ethnic-global-8-d2-b150-rsids.zip</span><span class="w"></span>
<span class="w">  </span><span class="nt">rfile</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">Multi-EthnicGlobal_D2_b150_rsids.txt</span><span class="w"></span>
</pre></div>
</div>
<p>Place your data into the <code class="docutils literal notranslate"><span class="pre">/Iliad/data/snp_array/idat/</span></code> directory.</p>
<p>Since this module is NOT the main snakefile, Snakemake will NOT automatically detect it without the <code class="docutils literal notranslate"><span class="pre">--snakefile</span></code> flag.
(Please make sure that your conda environment for Iliad is activated - <code class="docutils literal notranslate"><span class="pre">conda</span> <span class="pre">activate</span> <span class="pre">iliadEnv</span></code> or <code class="docutils literal notranslate"><span class="pre">mamba</span> <span class="pre">activate</span> <span class="pre">iliadEnv</span></code>)</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>snakemake --snakefile workflow/snpArray_Snakefile --cores <span class="m">1</span>
</pre></div>
</div>
<p>and combined with other user-specified snakemake flags such as <code class="docutils literal notranslate"><span class="pre">--cores</span></code>.</p>
<p>If you plan to use on a local machine or self-built server without a job scheduler the default command to run is the following:</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>snakemake -p --use-singularity --use-conda --snakefile workflow/snpArray_Snakefile --cores <span class="m">1</span> --jobs <span class="m">1</span> --default-resource<span class="o">=</span><span class="nv">mem_mb</span><span class="o">=</span><span class="m">10000</span> --latency-wait <span class="m">120</span>
</pre></div>
</div>
<p>However, there is a file included in the <code class="docutils literal notranslate"><span class="pre">Iliad</span></code> directory named - <code class="docutils literal notranslate"><span class="pre">snpArray-snakemake.sh</span></code> that will be useful in batch job submission.
Below is an example snakemake workflow submission in SLURM job scheduler.
Please read the shell variables at the top of the script and customize to your own paths and resource needs.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>sbatch snpArray-snakemake.sh
</pre></div>
</div>
<p>If you would like more in-depth information and descriptions, please continue to the next sections below.
Otherwise, you have completed the TL;DR setup section.</p>
</section>
<section id="information">
<h2>Information<a class="headerlink" href="#information" title="Permalink to this heading"></a></h2>
<hr class="docutils" />
<p>This tutorial introduces the genome-wide SNP array data processing module of the <a href="#id2"><span class="problematic" id="id3">Iliad_</span></a> workflow developed using Snakemake workflow language.
Please visit <a class="reference external" href="https://snakemake.readthedocs.io">Snakemake</a> for specific details. They also provide informational <a class="reference external" href="https://slides.com/johanneskoester/snakemake-tutorial">slides</a>. In general, though, each module is composed of rules. These rules define how output files are generated from input files while
automatically determining dependencies amongst the rules. A <code class="docutils literal notranslate"><span class="pre">DAG</span></code> (directed acyclic graph) of jobs will be built each time to account for all of the samples and jobs
that will executed either via job scheduler or local cores and will execute in parallel if multiple jobs are declared.
Because of the Snakemake workflow system design, the <strong>Iliad</strong> workflow is scalable from single core machines to HPC clusters with job schedulers.</p>
<p>The <strong>SNP array module</strong> is designed to process target data in your lab. Iliad is currently limited to <a class="reference external" href="https://support.illumina.com/">Illumina</a> microarray raw data processing and is configured for the
human genotyping Infinium Multi-Ethic Global-8 Kit (<a class="reference external" href="https://support.illumina.com/array/array_kits/infinium-multi-ethnic-global-8-kit.html">MEGA</a>).
We ensured no bioinformatics knowledge is needed to run this module with the help of external test runs performed on Google Cloud Platform (<a class="reference external" href="https://cloud.google.com/">GCP</a>).</p>
<p><strong>SNP Array Module Rule Graph</strong></p>
<a class="reference internal image-reference" href="../_images/snpArray-rulegraph.png"><img alt="../_images/snpArray-rulegraph.png" class="align-center" src="../_images/snpArray-rulegraph.png" style="width: 525px;" /></a>
<div class="toctree-wrapper compound">
</div>
</section>
<section id="background">
<h2>Background<a class="headerlink" href="#background" title="Permalink to this heading"></a></h2>
<hr class="docutils" />
<p>Genome-wide microarray data remains one of most widely used methods to obtain genotypic information on collected DNA samples, despite the
growing popularity and accessibility of genotyping by sequencing.
To make a comprehensive genomic pipeline,
we wanted to provide the means necessary for researchers to still access such a large body of data that remains important for many analyses.
<code class="docutils literal notranslate"><span class="pre">GWAS</span> <span class="pre">data</span></code> can be used in many more applications than gene identification, such as ancestry estimation,
historical population reconstruction, clinical genetic testing for diagnostic purposes, forensic analyses, and new method validation for sequencing data.</p>
<p>This module is currently limited to <a class="reference external" href="https://support.illumina.com/">Illumina</a> microarrays on the basis of the software tools and support and product file downloads.
It is configured to the <a class="reference external" href="https://support.illumina.com/array/array_kits/infinium-multi-ethnic-global-8-kit.html">MEGA</a> microarray, meaning download files are pointed to MEGA support files and product files.
It does possess the capability to be adapted to other microarrays. Pull requests and contributions are welcomed.</p>
</section>
<section id="basics">
<h2>Basics<a class="headerlink" href="#basics" title="Permalink to this heading"></a></h2>
<hr class="docutils" />
<p>The raw files from an Illumina sequencer are bead array files found in raw intensity data <code class="docutils literal notranslate"><span class="pre">.idat</span></code> format.
These <code class="docutils literal notranslate"><span class="pre">.idat</span></code> files are to be converted into Genotype Call <code class="docutils literal notranslate"><span class="pre">.gtc</span></code> files using <a class="reference external" href="https://support.illumina.com/downloads/iaap-genotyping-cli.html">iaap-cli</a> software. This software does have an
end-user license agreement (<a class="reference external" href="chrome-extension://efaidnbmnnnibpcajpcglclefindmkaj/https://support.illumina.com/content/dam/illumina-support/documents/downloads/software/iaap/Illumina%20Array%20Analysis%20Platform%20IAAP%201.1%20EULA.pdf">EULA</a>) and is not included or distributed by Iliad. If the user chooses to configure a download of the
program, it will be downloaded, independent from the Iliad repository distribution.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">.gtc</span></code> files are converted to a <code class="docutils literal notranslate"><span class="pre">.vcf</span></code> using <a class="reference external" href="https://samtools.github.io/bcftools/bcftools.html">bcftools</a> plug-in <a class="reference external" href="https://github.com/freeseek/gtc2vcf">gtc2vcf</a>.
This requires a reference genome assembly and Iliad downloads the user-configured reference genome fasta files.
Iliad is configured to download <em>Homo sapiens</em> GRCh38 release 104 as default.</p>
<p>Processing the <code class="docutils literal notranslate"><span class="pre">.vcf</span></code> is critical for realistic genetic data compatibility.
Custom python scripts are called in rules to rename unconventional loci names to standardized <code class="docutils literal notranslate"><span class="pre">rs</span> <span class="pre">IDs</span></code> using dbSNP files.
The default configuration file is set to download human_9606_b151_GRCh38p7 <code class="docutils literal notranslate"><span class="pre">All_20180418.vcf.gz</span></code>.</p>
<p>Once the <code class="docutils literal notranslate"><span class="pre">vcf</span></code> is processed, quality controls are performed based on the GenTrain and ClusterSep scores.
Default thresholds, along with other default workflow configurations, can be found in your path to the configuration file: <code class="docutils literal notranslate"><span class="pre">config/config.yaml</span></code>.</p>
<section id="in-depth-setup">
<h3>In-depth Setup<a class="headerlink" href="#in-depth-setup" title="Permalink to this heading"></a></h3>
<p>Once the <a class="reference external" href="https://iliad-readthedocs.readthedocs.io/en/latest/getting_started/installation.html">Installation</a> of Iliad and its two dependencies has been completed,
you will find your new working directory within the <code class="docutils literal notranslate"><span class="pre">PATH/TO/Iliad</span></code> folder.
Make sure your current working directory is in this cloned repo as stated in the installation.
If the repository is not cloned in that fashion, there is a chance that your direcory will be improperly named as <code class="docutils literal notranslate"><span class="pre">Iliad-main</span></code>.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span><span class="nb">cd</span> Iliad
</pre></div>
</div>
<p>In that working directory you will find there are a number of directories with files and code to run each of the module pipelines.</p>
<p><strong>FIRST</strong>,
there is a <code class="docutils literal notranslate"><span class="pre">data/snp_array/idat</span></code> directory with a <code class="docutils literal notranslate"><span class="pre">readme.md</span></code> file. You must place all of your <code class="docutils literal notranslate"><span class="pre">.idat</span></code> files in this folder.
There should be two <code class="docutils literal notranslate"><span class="pre">.idat</span></code> files for each sample: one green <code class="docutils literal notranslate"><span class="pre">_Grn.idat</span></code> and one red <code class="docutils literal notranslate"><span class="pre">_Red.idat</span></code>.</p>
<p><strong>SECOND</strong>,
there is a configuration file with some default parameters, however, you MUST at least change the <code class="docutils literal notranslate"><span class="pre">workdirPath</span></code> parameter to the appropriate
path leading up to and including <code class="docutils literal notranslate"><span class="pre">/Iliad</span></code> e.g. <code class="docutils literal notranslate"><span class="pre">/Path/To/Iliad/</span></code>. The configuration file is found in <code class="docutils literal notranslate"><span class="pre">config/config.yaml</span></code>.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">workdirPath</span><span class="p">:</span> <span class="o">/</span><span class="n">Path</span><span class="o">/</span><span class="n">To</span><span class="o">/</span><span class="n">Iliad</span><span class="o">/</span>
</pre></div>
</div>
<p>Some other parameters that are pre-set and you might consider changing to your project needs include:</p>
<ul class="simple">
<li><p>Homo sapiens GRCh38 release 104 reference genome</p></li>
</ul>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">ref</span><span class="p">:</span>
  <span class="n">species</span><span class="p">:</span> <span class="n">homo_sapiens</span>
  <span class="n">release</span><span class="p">:</span> <span class="mi">104</span>
  <span class="n">build</span><span class="p">:</span> <span class="n">GRCh38</span>
</pre></div>
</div>
<ul class="simple">
<li><p>Illumina <a class="reference external" href="https://support.illumina.com/array/array_kits/infinium-multi-ethnic-global-8-kit.html">MEGA</a> microarray GRCh38 support and product files</p></li>
</ul>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">urlProductFiles</span><span class="p">:</span>
  <span class="n">manifest</span><span class="p">:</span>  <span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">webdata</span><span class="o">.</span><span class="n">illumina</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">downloads</span><span class="o">/</span><span class="n">productfiles</span><span class="o">/</span><span class="n">multiethnic</span><span class="o">-</span><span class="k">global</span><span class="o">-</span><span class="mi">8</span><span class="o">/</span><span class="n">v1</span><span class="o">-</span><span class="mi">0</span><span class="o">/</span><span class="n">build38</span><span class="o">/</span><span class="n">multi</span><span class="o">-</span><span class="n">ethnic</span><span class="o">-</span><span class="k">global</span><span class="o">-</span><span class="mi">8</span><span class="o">-</span><span class="n">d2</span><span class="o">-</span><span class="n">bpm</span><span class="o">.</span><span class="n">zip</span>
  <span class="n">mzip</span><span class="p">:</span> <span class="n">multi</span><span class="o">-</span><span class="n">ethnic</span><span class="o">-</span><span class="k">global</span><span class="o">-</span><span class="mi">8</span><span class="o">-</span><span class="n">d2</span><span class="o">-</span><span class="n">bpm</span><span class="o">.</span><span class="n">zip</span>
  <span class="n">cluster</span><span class="p">:</span> <span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">webdata</span><span class="o">.</span><span class="n">illumina</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">downloads</span><span class="o">/</span><span class="n">productfiles</span><span class="o">/</span><span class="n">multiethnic</span><span class="o">-</span><span class="k">global</span><span class="o">-</span><span class="mi">8</span><span class="o">/</span><span class="n">v1</span><span class="o">-</span><span class="mi">0</span><span class="o">/</span><span class="n">infinium</span><span class="o">-</span><span class="n">multi</span><span class="o">-</span><span class="n">ethnic</span><span class="o">-</span><span class="k">global</span><span class="o">-</span><span class="mi">8</span><span class="o">-</span><span class="n">d1</span><span class="o">-</span><span class="n">cluster</span><span class="o">-</span><span class="n">file</span><span class="o">.</span><span class="n">zip</span>
  <span class="n">czip</span><span class="p">:</span> <span class="n">infinium</span><span class="o">-</span><span class="n">multi</span><span class="o">-</span><span class="n">ethnic</span><span class="o">-</span><span class="k">global</span><span class="o">-</span><span class="mi">8</span><span class="o">-</span><span class="n">d1</span><span class="o">-</span><span class="n">cluster</span><span class="o">-</span><span class="n">file</span><span class="o">.</span><span class="n">zip</span>

<span class="n">urlSupportFiles</span><span class="p">:</span>
  <span class="n">physicalGeneticCoordinates</span><span class="p">:</span> <span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">support</span><span class="o">.</span><span class="n">illumina</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">content</span><span class="o">/</span><span class="n">dam</span><span class="o">/</span><span class="n">illumina</span><span class="o">-</span><span class="n">support</span><span class="o">/</span><span class="n">documents</span><span class="o">/</span><span class="n">downloads</span><span class="o">/</span><span class="n">productfiles</span><span class="o">/</span><span class="n">multiethnic</span><span class="o">-</span><span class="k">global</span><span class="o">/</span><span class="n">multi</span><span class="o">-</span><span class="n">ethnic</span><span class="o">-</span><span class="k">global</span><span class="o">-</span><span class="mi">8</span><span class="o">-</span><span class="n">d2</span><span class="o">-</span><span class="n">physical</span><span class="o">-</span><span class="n">genetic</span><span class="o">-</span><span class="n">coordinates</span><span class="o">.</span><span class="n">zip</span>
  <span class="n">pzip</span><span class="p">:</span> <span class="n">multi</span><span class="o">-</span><span class="n">ethnic</span><span class="o">-</span><span class="k">global</span><span class="o">-</span><span class="mi">8</span><span class="o">-</span><span class="n">d2</span><span class="o">-</span><span class="n">physical</span><span class="o">-</span><span class="n">genetic</span><span class="o">-</span><span class="n">coordinates</span><span class="o">.</span><span class="n">zip</span>
  <span class="n">rsidConversion</span><span class="p">:</span> <span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">support</span><span class="o">.</span><span class="n">illumina</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">content</span><span class="o">/</span><span class="n">dam</span><span class="o">/</span><span class="n">illumina</span><span class="o">-</span><span class="n">support</span><span class="o">/</span><span class="n">documents</span><span class="o">/</span><span class="n">downloads</span><span class="o">/</span><span class="n">productfiles</span><span class="o">/</span><span class="n">multiethnic</span><span class="o">-</span><span class="k">global</span><span class="o">/</span><span class="n">multi</span><span class="o">-</span><span class="n">ethnic</span><span class="o">-</span><span class="k">global</span><span class="o">-</span><span class="mi">8</span><span class="o">-</span><span class="n">d2</span><span class="o">-</span><span class="n">b150</span><span class="o">-</span><span class="n">rsids</span><span class="o">.</span><span class="n">zip</span>
  <span class="n">rzip</span><span class="p">:</span> <span class="n">multi</span><span class="o">-</span><span class="n">ethnic</span><span class="o">-</span><span class="k">global</span><span class="o">-</span><span class="mi">8</span><span class="o">-</span><span class="n">d2</span><span class="o">-</span><span class="n">b150</span><span class="o">-</span><span class="n">rsids</span><span class="o">.</span><span class="n">zip</span>
  <span class="n">rfile</span><span class="p">:</span> <span class="n">Multi</span><span class="o">-</span><span class="n">EthnicGlobal_D2_b150_rsids</span><span class="o">.</span><span class="n">txt</span>
</pre></div>
</div>
<p><strong>THIRD</strong>,
each module pipeline has a specific <code class="docutils literal notranslate"><span class="pre">Snakefile</span></code>.
Snakemake will automatically detect the main snakefile, which is named excatly as such and found in the <code class="docutils literal notranslate"><span class="pre">workflow</span></code> directory: <code class="docutils literal notranslate"><span class="pre">workflow/Snakefile</span></code>.
Iliad reserves the main snakefile for the main module, specifically the raw sequence read data <a class="reference external" href="https://iliad-readthedocs.readthedocs.io/en/latest/tutorial/raw_sequence.html">module</a>.
This means the user must specify which <code class="docutils literal notranslate"><span class="pre">Snakefile</span></code> will be invoked with the following:</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>snakemake --snakefile workflow/snpArray_Snakefile
</pre></div>
</div>
<p>and combined with other user-specified snakemake flags, of course, like <code class="docutils literal notranslate"><span class="pre">--cores</span></code>.</p>
<p>In this module, the SNP Array Snakefile is also located in the <code class="docutils literal notranslate"><span class="pre">workflow</span></code> directory: <code class="docutils literal notranslate"><span class="pre">workflow/snpArray_Snakefile</span></code>.
Users must invoke this <code class="docutils literal notranslate"><span class="pre">snpArray_Snakefile</span></code> in order to run the workflow for This <strong>SNP ARRAY MODULE</strong>.</p>
<p>If you plan to use on a local machine or self-built server without a job scheduler the default command to run is the following:</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>snakemake -p --use-singularity --use-conda --cores <span class="m">1</span> --jobs <span class="m">1</span> --snakefile workflow/snpArray_Snakefile --default-resource<span class="o">=</span><span class="nv">mem_mb</span><span class="o">=</span><span class="m">10000</span> --latency-wait <span class="m">120</span>
</pre></div>
</div>
<p>However, there is a file included in the <code class="docutils literal notranslate"><span class="pre">Iliad</span></code> directory named - <code class="docutils literal notranslate"><span class="pre">snpArray-snakemake.sh</span></code> that will be useful in batch job submission.
Below is an example snakemake workflow submission in SLURM job scheduler.
Please read the shell variables at the top of the script and customize to your own paths and resource needs.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>sbatch snpArray-snakemake.sh
</pre></div>
</div>
</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="stored_sequence.html" class="btn btn-neutral float-left" title="Iliad Stored Sequence" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="liftover.html" class="btn btn-neutral float-right" title="Lift Over Variants - 37 / 38" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023-2023, Noah Herrick.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>